File: src/main/java/io/github/mschout/email/srs/provider/ShortCutSRSProvider.java
package io.github.mschout.email.srs.provider;

import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;
import io.github.mschout.email.srs.SRSAddress;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import lombok.Getter;
import lombok.experimental.SuperBuilder;

@SuperBuilder
public class ShortCutSRSProvider implements SRSProvider {
  @Getter
  private final List<String> secrets;

  private final Integer hashLength;

  private final Integer hashMinLength;

  @Getter
  protected final String separator;

  protected final String SRSSEP = "=";

  protected final String separators = "-+=";

  public ShortCutSRSProvider(List<String> secrets, int hashLength, int hashMinLength, String separator) {
    if (!separators.contains(separator)) throw new IllegalArgumentException("Initial separator must be = - or +, not " + separator);

    this.secrets = secrets;
    this.hashLength = hashLength;
    this.hashMinLength = hashMinLength;
    this.separator = separator;
  }

  protected String createHash(List<String> value) throws InvalidKeyException {
    return createHash(value, getSecret());
  }

  private String createHash(List<String> value, String secret) throws InvalidKeyException {
    final String algorithm = "HmacSHA1";

    try {
      SecretKeySpec secretKeySpec = new SecretKeySpec(secret.getBytes(), algorithm);
      Mac mac = Mac.getInstance(algorithm);
      mac.init(secretKeySpec);

      String data = String.join("", value).toLowerCase();

      return Base64.getEncoder().encodeToString(mac.doFinal(data.getBytes())).substring(0, hashLength);
    } catch (NoSuchAlgorithmException e) {
      // Really should never happen since we hard coded HmacSHA1
      throw new RuntimeException(e);
    }
  }

  protected boolean isHashInvalid(List<String> value, String hash) {
    if (hash.length() < hashMinLength) return false;

    List<String> validHashes = new ArrayList<>();

    for (String secret : secrets) {
      try {
        String candidate = createHash(value, secret);

        // If we got an exact match, bail out, the hash is valid.
        if (candidate.equals(hash)) return false;

        validHashes.add(candidate);
      } catch (InvalidKeyException e) {
        // invalid key = hash is invalid
        return true;
      }
    }

    // if we did not match any of the hashes exactly, try case-insensitive matching
    for (String candidate : validHashes) {
      if (candidate.equalsIgnoreCase(hash)) return false;
    }

    // nothing matched, its not valid
    return true;
  }

  protected boolean isSepChar(char ch) {
    return separators.indexOf(ch) != -1;
  }

  public boolean isSRS0(String address) {
    String prefix = SRSPrefix.SRS0;

    return (address.toUpperCase().startsWith(prefix) && isSepChar(address.charAt(prefix.length())));
  }

  public boolean isSRS1(String address) {
    String prefix = SRSPrefix.SRS1;
    return (address.toUpperCase().startsWith(prefix) && isSepChar(address.charAt(prefix.length())));
  }

  // Remove the SRS prefix tag plus the separator char that follows it.
  // If the address does not start with a SRS0 or SRS1 tag, returns the address as-is
  protected String removePrefix(String address) {
    if (isSRS1(address)) {
      return address.substring(SRSPrefix.SRS1.length() + 1);
    } else if (isSRS0(address)) {
      return address.substring(SRSPrefix.SRS0.length() + 1);
    } else {
      return address;
    }
  }

  @Override
  public String compile(String host, String user) throws InvalidKeyException {
    String timestamp = SRSTimestamp.generate();

    List<String> hashData = new ArrayList<>(3);
    hashData.add(timestamp);

    if (isSRS0(user)) {
      // This dulplicates effort in GuardedSRSProvider but makes this file work standalone
      user = removePrefix(user);

      List<String> addressIter = Splitter.on(SRSSEP).limit(4).splitToList(user);

      host = addressIter.get(2);
      user = addressIter.get(3);
    } else if (isSRS1(user)) {
      // This should never be hit in practice.  It would be bad.
      // Introduce compatibility with the guarded format?
      // tag, SRSHOST, hash, timestamp, host, user
      List<String> addressIter = Splitter.on(SRSSEP).limit(6).splitToList(user);
      host = addressIter.get(3);
      user = addressIter.get(4);
    }

    hashData.add(host);
    hashData.add(user);

    String hash = createHash(hashData);

    return SRSPrefix.SRS0 + getSeparator() + String.join(SRSSEP, hash, timestamp, host, user);
  }

  @Override
  public SRSAddress parse(String srsAddress) {
    if (!isSRS0(srsAddress)) throw new IllegalArgumentException(
      String.format("Reverse address %s does not start with %s=", srsAddress, SRSPrefix.SRS0)
    );

    String address = removePrefix(srsAddress);

    try {
      // The 4 here matches the number of fields we encoded above. If
      // there are more separators, then they belong in senduser anyway.
      Iterator<String> addressIter = Splitter.on(SRSSEP).limit(4).split(address).iterator();

      String hash = addressIter.next();
      String timestamp = addressIter.next();
      String host = addressIter.next();
      String user = addressIter.next();

      if (isHashInvalid(ImmutableList.of(timestamp, host, user), hash)) {
        throw new IllegalArgumentException("Invalid address hash: " + hash);
      }

      if (SRSTimestamp.isInvalid(timestamp)) throw new IllegalArgumentException("Invalid timestamp");

      return new SRSAddress(SRSPrefix.SRS0, host, user, hash);
    } catch (NoSuchElementException e) {
      throw new IllegalArgumentException("Invalid SRS Address: " + srsAddress);
    }
  }

  @Override
  public String getSecret() {
    return secrets.get(0);
  }
}

File: src/main/kotlin/io/github/mschout/email/srs/provider/ShortCutSRSProvider.kt
package io.github.mschout.email.srs.provider

import io.github.mschout.email.srs.SRSAddress
import java.security.InvalidKeyException
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Implementation of the SRSProvider interface for handling Sender Rewriting Scheme (SRS). This
 * class provides functionalities to encode and decode email addresses for forwarding while
 * maintaining SPF (Sender Policy Framework) compliance.
 *
 * @param secrets A list of secrets used for generating and validating hashes.
 * @param hashLength The length of the hash to be generated.
 * @param hashMinLength The minimum valid length of a hash.
 * @param separator A character used as a delimiter in the generated SRS address. Must be one of the
 *   valid separators: `=`, `-`, or `+`.
 * @property secret Retrieves the primary secret from the list of provided secrets.
 * @property separator The specified delimiter used within the SRS addresses.
 * @constructor Creates a ShortCutSRSProvider instance with the specified configuration.
 * @throws IllegalArgumentException If the provided separator is not a valid character.
 */
open class ShortCutSRSProvider(
    val secrets: List<String>,
    private val hashLength: Int,
    private val hashMinLength: Int,
    override val separator: String,
) : SRSProvider {

  protected val SRSSEP = "="
  protected val separators = "-+="

  init {
    require(separator in separators.map { it.toString() }) {
      "Initial separator must be = - or +, not $separator"
    }
  }

  override val secret: String
    get() = secrets[0]

  @Throws(InvalidKeyException::class)
  protected fun createHash(value: List<String>): String = createHash(value, secret)

  private fun createHash(value: List<String>, secret: String): String {
    val algorithm = "HmacSHA1"
    val secretKeySpec = SecretKeySpec(secret.toByteArray(), algorithm)
    val mac = Mac.getInstance(algorithm)
    mac.init(secretKeySpec)

    val data = value.joinToString("").lowercase()
    return Base64.getEncoder()
        .encodeToString(mac.doFinal(data.toByteArray()))
        .substring(0, hashLength)
  }

  protected fun isHashInvalid(value: List<String>, hash: String): Boolean {
    if (hash.length < hashMinLength) return false

    val validHashes = mutableListOf<String>()

    for (secret in secrets) {
      try {
        val candidate = createHash(value, secret)
        if (candidate == hash) return false
        validHashes.add(candidate)
      } catch (_: InvalidKeyException) {
        return true
      }
    }

    for (candidate in validHashes) {
      if (candidate.equals(hash, ignoreCase = true)) return false
    }

    return true
  }

  protected fun isSepChar(ch: Char): Boolean = ch in separators

  override fun isSRS0(address: String): Boolean {
    return address.uppercase().startsWith(SRS0_PREFIX) && isSepChar(address[SRS0_PREFIX.length])
  }

  override fun isSRS1(address: String): Boolean {
    return address.uppercase().startsWith(SRS1_PREFIX) && isSepChar(address[SRS1_PREFIX.length])
  }

  protected fun removePrefix(address: String): String =
      when {
        isSRS1(address) -> address.substring(SRS1_PREFIX.length + 1)
        isSRS0(address) -> address.substring(SRS0_PREFIX.length + 1)
        else -> address
      }

  @Throws(InvalidKeyException::class)
  override fun compile(host: String, user: String): String {
    var actualHost = host
    var actualUser = user
    val timestamp = SRSTimestamp.generate()

    if (isSRS0(actualUser)) {
      actualUser = removePrefix(actualUser)
      val parts = actualUser.split(SRSSEP, limit = 4)
      actualHost = parts[2]
      actualUser = parts[3]
    } else if (isSRS1(actualUser)) {
      val parts = actualUser.split(SRSSEP, limit = 6)
      actualHost = parts[3]
      actualUser = parts[4]
    }

    val hash = createHash(listOf(timestamp, actualHost, actualUser))

    return SRS0_PREFIX +
        separator +
        listOf(hash, timestamp, actualHost, actualUser).joinToString(SRSSEP)
  }

  override fun parse(srsAddress: String): SRSAddress {
    require(isSRS0(srsAddress)) { "Reverse address $srsAddress does not start with $SRS0_PREFIX=" }

    val address = removePrefix(srsAddress)

    try {
      val parts = address.split(SRSSEP, limit = 4)
      val hash = parts[0]
      val timestamp = parts[1]
      val host = parts[2]
      val user = parts[3]

      if (isHashInvalid(listOf(timestamp, host, user), hash)) {
        throw IllegalArgumentException("Invalid address hash: $hash")
      }

      if (SRSTimestamp.isInvalid(timestamp)) {
        throw IllegalArgumentException("Invalid timestamp")
      }

      return SRSAddress(SRS0_PREFIX, host, user, hash)
    } catch (e: IndexOutOfBoundsException) {
      throw IllegalArgumentException("Invalid SRS Address: $srsAddress")
    }
  }
}
